src/recovery/mod.rs

```diff
@@ -0,0 +1,282 @@
+
+        SqliteRecoveryStore { rt, pool }
+    }
+}
+
+impl RecoveryStore for SqliteRecoveryStore {
+    fn for_step(&self, step_id: &str) -> Box<dyn StepRecovery<u64, TdPyAny, TdPyAny>> {
+        let future = query(
+            "CREATE TABLE IF NOT EXISTS states(epoch INTEGER, step_id, key, state, PRIMARY KEY (epoch, step_id, key));"
+        ).execute(&self.pool);
+        self.rt.block_on(future).unwrap();
+
+        Box::new(SqliteStepRecovery {
+            step_id: step_id.to_string(),
+            rt: self.rt.clone(),
+            pool: self.pool.clone(), // Said to be cheap.
+        })
+    }
+}
+
+struct SqliteStepRecovery {
+    step_id: String,
+    rt: SendWrapper<Rc<Runtime>>,
+    pool: Pool<Sqlite>,
+}
+
+impl StepRecovery<u64, TdPyAny, TdPyAny> for SqliteStepRecovery {
+    fn recover_last(&self, current_epoch: &u64, key: &TdPyAny) -> Option<TdPyAny> {
+        let current_epoch_int =
+            i64::try_from(*current_epoch).expect("Epoch too big to fit into Sqlite int");
+        let key_string: String =
+            Python::with_gil(|py| key.extract(py)).expect("Key cannot be cast to string");
+
+        let future = query("SELECT state FROM states WHERE step_id = ?1 AND key = ?2 AND epoch < ?3 ORDER BY epoch DESC LIMIT 1")
+            .bind(&self.step_id)
+            .bind(key_string)
+            .bind(current_epoch_int)
+            .map(|row: SqliteRow| {
+                let state_pickled: &[u8] = row.get(0);
+                Python::with_gil(|py| {
+                    let pickle = py.import("dill")?;
+                    Ok(pickle.call_method1("loads", (state_pickled, ))?.into())
+                })
+            })
+            .fetch_optional(&self.pool);
+        let state = self
+            .rt
+            .block_on(future)
+            .unwrap()
+            .map(|r: Result<TdPyAny, PyErr>| r.expect("Error unpickling state"));
+
+        debug!(
+            "sqlite state recovered step_id={} key={key:?}:state={state:?}@epoch={current_epoch}",
+            self.step_id
+        );
+        state
+    }
+
+    fn save_complete(&self, completed_epoch: &u64, key: &TdPyAny, state: &Option<TdPyAny>) -> () {
+        let completed_epoch_int =
+            i64::try_from(*completed_epoch).expect("Epoch too big to fit into Sqlite int");
+        let key_string: String =
+            Python::with_gil(|py| key.extract(py)).expect("Key cannot be cast to string");
+        let state_pickled: Option<Vec<u8>> = state.as_ref().map(|state| {
+            Python::with_gil(|py| -> Result<Vec<u8>, PyErr> {
+                let pickle = py.import("dill")?;
+                Ok(pickle.call_method1("dumps", (state,))?.extract()?)
+            })
+            .expect("Error pickling state")
+        });
+
+        let future = query("INSERT INTO states (epoch, step_id, key, state) VALUES (?1, ?2, ?3, ?4) ON CONFLICT (epoch, step_id, key) DO UPDATE SET state = EXCLUDED.state")
```

Comment:
I'm not totally clear on the context of your question. The high-level purpose of the `ON CONFLICT` clause is so that if a second dataflow execution re-calculates the state for some key at some epoch (which very easily could happen if we're restarting and one worker is ahead in an epoch than another; we have to go back to the oldest finalized epoch), the re-calculation is saved _overwriting_ the old value. I think overwriting is less surprising than ignoring since the resulting state on the second pass; it feels more correct that the results of the most recent input data should be used.

You might argue that we should hard error on this situation (or have a toggle to), but then we'd have to have a surefire way to know which recovery data to delete to allow a graceful restart (since some workers might be ahead of others). I'd guess that this would mean "delete everything after the latest finalized / sealed epoch" (once we implement some way of calculating that), but I'm not totally sure that's correct in the general case (especially with delays). This seems easier for now.